doc/security.rst

```diff
@@ -0,0 +1,184 @@
+##################################################################
+
+Several technologies which can be used to increase the security of
+actual products are already enabled in the IoT Reference OS Kit. Which
+of these are suitable for a given product depends on the design goals
+for that product.
+
+This document summarizes the available technologies and where to find
+further information. Developers are encouraged to read the
+``.bbclass`` and ``.bb`` files to understand how the different
+technologies are integrated and how to configure or use them.
+
+Integrity Protection
+====================
+
+The goal of integrity protection is to ensure that a device boots up
+using the software and configuration provided by the vendor. Attackers
+may try to influence that both when the device is offline (for
+example, by mounting and modifying the filesystem of the device) and
+when it is online (by running malicious code and writing to the
+filesystem).
+
+Read-only Root Filesystem
+-------------------------
+
+Protecting a read-only root filesystem is possible with
+`dm-verity`_. In this mode, the entire root partition is signed at
+build time with a secret private key. The ``refkit-initramfs``
+verifies that signature using a public key before enabling access to
+the partition. Then at runtime, dm-verity detects modified blocks
+and rejects access to them, which shows up as read errors at the block
+level.
+
+.. _dm-verity: https://source.android.com/security/verifiedboot/
+
+In other words, modifications are detected reliably and
+immediately. Because the secret key is not on the device, not even the
+kernel can modify the protected partition.
+
+How to design a product such that it works with a read-only rootfs is
+out of scope for this document.
+
+To use dm-verity, an image must have the ``read-only-rootfs
+dm-verity`` ``IMAGE_FEATURES`` and be derived from the
+:file:`meta-refkit/classes/refkit-image.bbclass`. The
+:file:`meta-refkit/recipes-images/refkit-installer-image.bb` is an
+example for an image that is built with dm-verity enabled.
+
+In development mode (= including :file:`refkit-development.inc`), a
+key provided in :file:`meta-refkit/files/dm-verity/private.pem` is
+used for signing, which is obviously not secure.
+
+Real products must use their own key. Those can be created for
+example with::
+
+  openssl genrsa -aes128 -passout pass:<your key password> -out private.pem 4096
+
+Then during the build, ``REFKIT_DMVERITY_PRIVATE_KEY`` must be set to
+the full path of the resulting :file:`private.pem` and
+``REFKIT_DMVERITY_PASSWORD`` must be ``pass:<your key password>`` or
+something else that tells openssl how to obtain the key password (see
+``-passin`` in ``man openssl``).
+
+A real product also needs to ensure that the device boots with an
+initramfs which really activates dm-verity.
+:file:`meta-refkit/recipes-image/images/initramfs-framework-refkit-dm-verity.bb`
+in ``refkit-initramfs`` does that when ``dmverity`` is part of the
```

Comment:
The documentation refers to this line:

          if [ -e "$bootparam_root" ] && [ -e "$bootparam_dmverity" ]; then

It's the dmverity boot parameter which must be present to trigger the code, which must be set when building the image. Would it be better to say:

> A real product also needs to ensure that its image really activates dm-verity. This is done by
> booting with the ``dmverity`` boot parameter, which tells :file:`meta-refkit/recipes-image/images/initramfs-framework-refkit-dm-verity.bb`
> in ``refkit-initramfs`` that dm-verity is required. :file:`refkit-image.bbclass` automatically adds that boot parameter when
> building an image with dm-verity enabled. Secure Boot is required to prevent attackers from replacing the initramfs 
> and/or changing boot parameters.